src/libraries/System.Security.Cryptography/docs/X509CertificateLoader.SecurityDesign.md

```diff
@@ -0,0 +1,335 @@
+
+Whenever a PKCS#12 PFX load is being rejected for exceeding the work limit, an instance of the dedicated exception type `Pkcs12LoadLimitExceededException` (which derives from `CryptographicException`) is thrown.
+This exception will contain the name of the property whose limit was exceeded in its message (or, if multiple limits were exceeded simultaneously, the limit deemed most appropriate), but this is purely for diagnostic purposes.
+
+### RFC 7292 Section 4, MacData
+
+The MacData structure contains a field, `iterations`, that specifies the number of iterations to use when running the Key Derivation Function (KDF) to turn the user-provided password into the MAC key.
+Callers can limit the number of iterations via the `MacIterationLimit` property, which defaults (as of .NET 9) to 300,000 iterations.
+
+The MacData structure also, via the `mac` field, can specify the hash algorithm underlying the PKCS12 KDF algorithm.
+While the performance of the KDF utilizing SHA-1 may be different than when utilizing SHA-2-512, `X509CertificateLoader` considers them be be close enough in their performance characteristics that no filtering or weighting is done on the algorithm.
+
+### RFC 7292 Section 4.1, AuthenticatedSafe
+
+The PKCS#12 PFX AuthenticatedSafe top-level structure allows for the SafeContents values it contains to have one of three confidentiality modes: plaintext (PKCS#7/CMS Data), password (PKCS#7/CMS EncryptedData), and public key (PKCS#7/CMS EnvelopedData).
+Public-key decryption requires a different API shape than password-based decryption, and as such is not currently supported.
+
+The password-based confidentiality mode allows specifying a symmetric encryption algorithm, a KDF algorithm, and the hash algorithm that underlies the chosen KDF algorithm.
+As with MacData, it is acknowledged that not all algorithms have the exact same performance characteristics, but as all of the choices are _O(n)_ with respect to the amount of data to process, and _n_ is bounded by the length of the payload, there are no controls to restrict or weight any algorithm choices.
+
+Password-based confidentiality also specifies the number of iterations to run for the chosen KDF.
+Callers can limit the number of iterations for the KDF on a single encrypted SafeContents via the `IndividualKdfIterationLimit` property (which .NET 9 defaults to 300,000).
+As there can be more than one password-based SafeContents, callers can limit the total number of iterations across all KDFs via the `TotalKdfIterationLimit` property (which .NET 9 defaults to 1,000,000).
+The `TotalKdfIterationLimit` property is also shared with ShroudedKeyBag entries, there are not separate controls for totals for SafeContents vs ShroudedKeyBags.  If either the `IndividualKdfIterationLimit` or the `TotalKdfIterationLimit` would be exceeded when decrypting a SafeContents value, an exception is thrown BEFORE doing the work associated with the limit.
```

Comment:
It may be worth noting that some work might be done before limits are exceeded. We generally cannot guarantee all of the limits are known up-front. A ShroudedKeyBag inside an encrypted AuthSafe means we need to do work to decrypt the authsafe before we know the work involved with the encrypted key bags.

Reply:
If you have some suggestion as to how to word that in, I'm happy to take it.  To me, it's obvious by "would be exceeded **when decrypting**", indicating that it is evaluated at each individual decryption operation.

Reply:
Eh, okay, it's probably fine as is and I am nitpicking.